### PR TITLE
Add support for rocThrust in sort when using HIP

### PR DESCRIPTION
**Pull request by Rombur**
This PR add support for rocThrust. Unlike Thrust, rocThrust is not always installed with HIP. For instance, the docker images do not contain rocThrust unless you use the `-complete` tag. I will note that the 3.0 release of rocThrust is not backward compatible with the 2.X  series. This is not an issue for us because `sort` is the same across version but it is problematic if we decide to require a minimum version of rocThrust because then the 3.0 version will be rejected. Since we are not affected by the break of backward compatibility, I've decided to forego the minimum version requirements. In practice,  this means that the minimum version of rocThrust is the minimum version that is compatible with ROCm 5.2

Once this PR is merged, I'll open a PR to document the new option that enables/disables rocThrust.

### CODE REVIEW

File: cmake/KokkosCore_config.h.in [hunk 1]
```diff
@@ -62,6 +62,7 @@
 #cmakedefine KOKKOS_ENABLE_LIBDL
 #cmakedefine KOKKOS_ENABLE_LIBQUADMATH
 #cmakedefine KOKKOS_ENABLE_ONEDPL
+#cmakedefine KOKKOS_ENABLE_ROCTHRUST
```
Comment (dalg24) [inline cmake/KokkosCore_config.h.in:65]: You might want to add that setting to `HIP::print_configuration()`

@masterleinad same comment about `KOKKOS_ENABLE_ONEDPL` and `SYCL`

Comment (masterleinad) [reply cmake/KokkosCore_config.h.in:65]: Let me do that after #6758 is merged.

File: containers/src/CMakeLists.txt [hunk 2]
```diff
@@ -28,6 +28,9 @@ KOKKOS_LIB_INCLUDE_DIRECTORIES(kokkoscontainers
   ${CMAKE_CURRENT_BINARY_DIR}
   ${CMAKE_CURRENT_SOURCE_DIR}
 )
+
+KOKKOS_LINK_TPL(kokkoscontainers PUBLIC ROCTHRUST)
```
Comment (dalg24) [inline containers/src/CMakeLists.txt:32]: Why link containers?

File: cmake/Modules/FindTPLROCTHRUST.cmake [hunk 3]
```diff
@@ -0,0 +1,2 @@
+FIND_PACKAGE(rocthrust REQUIRED)
+KOKKOS_CREATE_IMPORTED_TPL(ROCTHRUST INTERFACE LINK_LIBRARIES roc::rocthrust)
```
Comment (dalg24) [inline cmake/Modules/FindTPLROCTHRUST.cmake:2]: Do we still have the https://github.com/arborx/ArborX/pull/652 issue where the exported rocThrust package potentially adds flags to compile for all GPU architecture it knows about?

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

File: containers/src/CMakeLists.txt (hunk 2 above)
Comment (Rombur) [reply containers/src/CMakeLists.txt:32]: My bad, I meant to add this in algorithm.

File: cmake/Modules/FindTPLROCTHRUST.cmake (hunk 3 above)
Comment (Rombur) [reply cmake/Modules/FindTPLROCTHRUST.cmake:2]: At least for the version shipped with ROCm 5.7, it works fine. I see only one architecture.

File: cmake/Modules/FindTPLROCTHRUST.cmake [hunk 4]
```diff
@@ -0,0 +1,2 @@
+FIND_PACKAGE(rocthrust REQUIRED)
```
Comment (dalg24) [inline cmake/Modules/FindTPLROCTHRUST.cmake:1]: Do we have a way for users to say I don't have rocThrust, I understand I am potentially leaving performance on the table but I want configuration to proceed without error?
Would `CMAKE_DISABLE_FIND_PACKAGE_<PackageName>` work here?
Or is the intent that they explicitly pass `-DKokkos_ENABLE_ROCTHRUST=OFF`

Comment (Rombur) [reply cmake/Modules/FindTPLROCTHRUST.cmake:1]: Yes, you should use `Kokkos_ENABLE_ROCTHRUST=OFF` to disable rocThrust.

File: cmake/Modules/FindTPLROCTHRUST.cmake (hunk 3 above)
Comment (dalg24) [reply cmake/Modules/FindTPLROCTHRUST.cmake:2]: We need to find out for sure because IIRC compile time blows up

**Comment by stanmoore1** [comment]
This would be useful for LAMMPS; `Kokkos::BinSort` isn't cutting it (https://github.com/lammps/lammps/issues/4075) at large scale on Frontier.

**Review by dalg24 — COMMENTED** [review]
Just realized my comment was pending.  Sorry about that.
I had mentioned the issue to Bruno in private.

**Comment by stanmoore1** [comment]
This will be in 4.3 correct? Would be really nice if it supported `sort_by_key` too...

**Comment by dalg24** [comment]
> This will be in 4.3 correct? Would be really nice if it supported `sort_by_key` too...

This is the plan yes.  Bruno is swamped with something else but will get back to it next week.  https://github.com/kokkos/kokkos/pull/6793#discussion_r1487197538 needs to be addressed and then we will merge.

**Comment by aprokop** [comment]
Rebased on `develop` to allow integration of rocThrust's `sort_by_key` in this PR.

**Comment by masterleinad** [comment]
Make sure to take https://github.com/kokkos/kokkos/pull/6856/files into account and define
```
template <class Layout>
inline constexpr bool sort_on_device_v<Kokkos::HIP, Layout> = true;
```

**Comment by Rombur** [comment]
This PR is to add support for rocThrust. `sort_by_key` will be its own PR.

**Review by masterleinad — APPROVED** [review]

**Review by dalg24 — APPROVED** [review]

**Pull request merged**